### ISSUE TITLE
[ENH] Add Generalized Score Functions using RKHS Regression for Nonlinear Causal Discovery

### ISSUE THREAD

**Issue by XAheli** [pgmpy/pgmpy#2502]
### **Is your feature request related to a problem? Please describe.**

Implementation of two generalized score functions based on RKHS (Reproducing Kernel Hilbert Space) regression that can handle nonlinear causal mechanisms, arbitrary data distributions and mixed-type data without assuming specific parametric models.

#### Motivation

Current score based methods in pgmpy (BIC, BDeu, BDs, K2, etc) make strong assumptions about the underlying causal mechanisms and data distributions:

- **BIC/AIC-Gauss**: Assumes linear-Gaussian relationships
- **BDeu/BDs/K2**: Designed for discrete data only
- **BIC/AIC-CondGauss**: Limited to specific mixed models

**These fail when**:

1. **Spurious edges**: Linear assumptions miss that variables are conditionally independent through nonlinear relationships
   ```python
   # Example: X₁ → X₂ → X₃ with quadratic relationships
   X₁ = E₁
   X₂ = 0.8(X₁ + X₁²) + E₂
   X₃ = 0.8(X₂ + X₂²) + E₃
   # BIC adds spurious edge X₁ → X₃ because partial correlation ≠ 0
   ```

2. **Missing edges**: Uncorrelated but dependent variables
   ```python
   # Example: Nonlinear dependency
   X₁ = E₁
   X₂ = (sin(X₁) + E₂)²
   # BIC misses X₁ → X₂ because correlation ≈ 0
   ```

### **Describe the solution you'd like**

Implementation of two nonparametric score functions from [Huang et al. (KDD 2018)](https://dl.acm.org/doi/10.1145/3219819.3220104)

#### 1. **Cross-Validated Likelihood Score** (`RKHSCVLikelihood`)

Uses kernel ridge regression in RKHS with cross-validation:

```
S_CV(Xᵢ, Parents) = (1/Q) Σ_{q=1}^Q ℓ(F̂ᵢ^(q) | D_test^(q))
```

Where:
- Split data into Q folds (K-fold or Monte Carlo CV)
- Train RKHS regression on each training fold
- Evaluate log-likelihood on corresponding test fold
- Average across folds

**Properties**
- Locally consistent (finds correct dependencies asymptotically)
- Works with existing GES/HillClimb via local score changes
- Not score equivalent (different DAGs in same MEC have different scores)

#### 2. **Marginal Likelihood Score** (`RKHSMarginalLikelihood`)

Uses Gaussian kernel formulation with hyperparameter optimization:

```
S_M(Xᵢ, Parents) = -½ trace{K̃_Xᵢ(K̃_Parents + σ̂ᵢ²I)⁻¹K̃_Xᵢ}
                   - (n/2)log|K̃_Parents + σ̂ᵢ²I| - (n²/2)log(2π)
```

Where σ̂ᵢ² is optimized via gradient methods.

**Properties**
- Locally consistent (via Laplace approximation)
- Works with GES/HillClimb
- Often faster than CV likelihood

#### Technical Details

**Core Method**: Regression in RKHS
- Map variables to infinite-dimensional Hilbert space: `φ_X(X) = F(Z) + U`
- Use empirical feature map (kernel trick) for finite samples
- Apply kernel ridge regression with Gaussian kernel

**Conditional Independence**: Characterized by cross-covariance operators
```
Σ_XX|Z - Σ_XX|[Y,Z] = 0  ⟺  X ⊥ Y|Z
```

**Key Insight**: Causal structure learning becomes model selection for RKHS regression tasks

**Default Hyperparameters** (from paper)
- Kernel: Gaussian/RBF
- Kernel width: 2 × median pairwise distance
- Cross-validation: 10-fold
- Regularization: λ = O(n^(-b/(bc+1)))

### **Implementation Plan**

**Module**: `pgmpy/estimators/StructureScore.py`

Add two new classes inheriting from `StructureScore`

```python
class RKHSCVLikelihood(StructureScore):
    """
    Cross-Validated Likelihood score using RKHS regression.

    Handles nonlinear causal mechanisms, arbitrary distributions,
    and mixed-type data without parametric assumptions.

    References
    ----------
    Huang, B., Zhang, K., Lin, Y., Schölkopf, B., & Glymour, C. (2018).
    Generalized Score Functions for Causal Discovery. KDD 2018.
    https://doi.org/10.1145/3219819.3220104
    """

    def __init__(self, data, kernel='rbf', kernel_width=None,
                 n_folds=10, lambda_reg='auto', **kwargs):
        """
        Parameters
        ----------
        data : pd.DataFrame
            Dataset
        kernel : str, default='rbf'
            Kernel type (currently only 'rbf' supported)
        kernel_width : float, optional
            Kernel bandwidth. If None, set to 2 × median distance
        n_folds : int, default=10
            Number of CV folds
        lambda_reg : float or 'auto', default='auto'
            Regularization parameter
        """
        super().__init__(data, **kwargs)
        self.kernel = kernel
        self.kernel_width = kernel_width
        self.n_folds = n_folds
        self.lambda_reg = lambda_reg

    def local_score(self, variable, parents):
        """
        Compute local CV likelihood score for variable given parents.

        Implements Equations 7-8 from Huang et al. (2018).
        """
        # Implementation here
        pass

    def _compute_kernel_matrix(self, X, Y=None):
        """Compute RBF kernel matrix K(X, Y)"""
        pass

    def _kernel_ridge_regression(self, K_X, K_Z, lambda_reg):
        """Perform kernel ridge regression"""
        pass


class RKHSMarginalLikelihood(StructureScore):
    """
    Marginal Likelihood score using RKHS regression.

    References
    ----------
    Huang, B., Zhang, K., Lin, Y., Schölkopf, B., & Glymour, C. (2018).
    Generalized Score Functions for Causal Discovery. KDD 2018.
    https://doi.org/10.1145/3219819.3220104
    """

    def __init__(self, data, kernel='rbf', kernel_width=None,
                 lambda_reg='auto', optimize_sigma=True, **kwargs):
        """
        Parameters
        ----------
        data : pd.DataFrame
            Dataset
        kernel : str, default='rbf'
            Kernel type
        kernel_width : float, optional
            Kernel bandwidth
        lambda_reg : float or 'auto'
            Regularization parameter
        optimize_sigma : bool, default=True
            Whether to optimize noise variance σ²
        """
        super().__init__(data, **kwargs)
        self.kernel = kernel
        self.kernel_width = kernel_width
        self.lambda_reg = lambda_reg
        self.optimize_sigma = optimize_sigma

    def local_score(self, variable, parents):
        """
        Compute local marginal likelihood score.

        Implements Equation 9 from Huang et al. (2018).
        """
        # Implementation here
        pass

    def _optimize_sigma(self, K_X, K_Parents):
        """Optimize noise variance hyperparameter"""
        pass
```

**Update `get_scoring_method()`** in `StructureScore.py`:

```python
def get_scoring_method(...):
    available_methods = {
        "continuous": {
            "bic-g": BICGauss,
            "ll-g": LogLikelihoodGauss,
            "aic-g": AICGauss,
            "rkhs-cv": RKHSCVLikelihood,        # NEW
            "rkhs-marginal": RKHSMarginalLikelihood,  # NEW
        },
        # ... existing methods
    }
```

**Works automatically with**:
- `GES` (Greedy Equivalence Search)
- `HillClimbSearch`
- Any algorithm using `StructureScore` interface

### **Expected Performance based on paper's experimental results**

| Scenario | Performance vs BIC/BDeu |
|----------|------------------------|
| **Nonlinear mechanisms** | Much better (handles sin, exp, polynomial) |
| **Dense graphs** (density > 0.4) | Better (10-20% higher F1) |
| **Small samples** (n=500) | Better (more robust) |
| **Multi-dimensional variables** | Much better (native support) |
| **Mixed data types** | Better (unified framework) |
| **Linear-Gaussian data** | ≈ Similar (slight overhead) |
| **Large discrete networks** | ≈ Slightly worse than BDeu |

**Benchmark Tests**
- Reproduce key results from paper 
- Test on CHILD and SACHS networks
- Compare with BIC, BDeu on various data types

### References

- Huang, B., Zhang, K., Lin, Y., Schölkopf, B., & Glymour, C. (2018). Generalized Score Functions for Causal Discovery. In KDD '18. ACM. https://doi.org/10.1145/3219819.3220104
- https://github.com/Biwei-Huang/Generalized-Score-Functions-for-Causal-Discovery

**Comment by XAheli** [comment]
cc @ankurankan

**Comment by XAheli** [comment]
https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.pairwise_kernels.html

Ref: versatility in choosing kernel methods from sklearn

**Comment by XAheli** [comment]
> This issue has been open for 1 month with no activity and is now marked as stale.

WiP
